### PR TITLE
refactor: `Skeleton.toDB` method

### PR DESCRIPTION
**Pull request by ArneGudermann**

### CODE REVIEW

File: src/viur/core/skeleton.py [hunk 1]
```diff
@@ -873,185 +873,176 @@ def toDB(cls, skelValues: SkeletonInstance, update_relations: bool = True, **kwa
...
+                        if lock_db_obj := db.Get(new_lock_key):
                             # There's already a lock for that value, check if we hold it
-                            if lockObj["references"] != dbObj.key.id_or_name:
+                            if lock_db_obj["references"] != db_obj.key.id_or_name:
                                 # This value has already been claimed, and not by us
                                 # TODO: Use a custom exception class which is catchable with an try/except
-                                raise ValueError(
-                                    "The unique value '%s' of bone '%s' has been recently claimed!" %
-                                    (skelValues[key], key))
+                                raise ValueError(f"The unique value '{skel[bone_name]}' of bone '{bone_name}' has been recently claimed!")
```
Comment (phorward) [inline src/viur/core/skeleton.py:954]:
```suggestion
                                raise ValueError(
                                    f"The unique value '{skel[bone_name]}' of bone '{bone_name}' "
                                    "has been recently claimed!"
                                )
```

**Review by phorward — APPROVED** [review]
Hello @ArneGudermann, let's check it out.
Some parts of this PR will conflict with #1025

**Comment by phorward** [comment]
@sveneberth can you please make a review as well on this?

File: src/viur/core/skeleton.py [hunk 2]
```diff
@@ -873,185 +873,179 @@ def toDB(cls, skelValues: SkeletonInstance, update_relations: bool = True, **kwa
...
+            db_key = write_skel["key"]
+            skel = write_skel.skeletonCls()
 
+            blob_list = set()
+            change_list = []
+            old_copy = {}
             # Load the current values from Datastore or create a new, empty db.Entity
-            if not dbKey:
+            is_add = not bool(db_key)
+            if not db_key:
```
Comment (sveneberth) [inline src/viur/core/skeleton.py:893]: IMO it's easier to understand if you would switch the if/else blocks and remove the `not`.

File: src/viur/core/skeleton.py [hunk 3]
```diff
@@ -873,185 +873,179 @@ def toDB(cls, skelValues: SkeletonInstance, update_relations: bool = True, **kwa
...
+
+            if not isinstance(db_obj["viur"].get("viurCurrentSeoKeys"), dict):
+                db_obj["viur"]["viurCurrentSeoKeys"] = {}
+            if current_seo_keys := skel.getCurrentSEOKeys():
                 # Convert to lower-case and remove certain characters
-                for lang, value in list(currentSeoKeys.items()):
-                    value = value.lower()
-                    value = value.translate(Skeleton.__seo_key_trans).strip()
-                    currentSeoKeys[lang] = value
+                for lang, value in list(current_seo_keys.items()):
```
Comment (sveneberth) [inline src/viur/core/skeleton.py:991]: I think the list case is not necessary here
```suggestion
                for lang, value in current_seo_keys.items():
```

File: src/viur/core/skeleton.py [hunk 4]
```diff
@@ -1061,75 +1055,75 @@ def fixDotNames(entity):
...
+            if None in blob_list:
+                msg = f"None is not valid in {blob_list=}"
+                logging.error(msg)
+                raise ValueError(msg)
+
+            if not is_add and (old_blob_lock_obj := db.Get(db.Key("viur-blob-locks", db_key.id_or_name))):
+                removed_blobs = set(old_blob_lock_obj.get("active_blob_references", [])) - blob_list
+                old_blob_lock_obj["active_blob_references"] = list(blob_list)
+                if old_blob_lock_obj["old_blob_references"] is None:
+                    old_blob_lock_obj["old_blob_references"] = [removed_blob for removed_blob in removed_blobs]
```
Comment (sveneberth) [inline src/viur/core/skeleton.py:1076]:
```suggestion
                    old_blob_lock_obj["old_blob_references"] = list(removed_blobs)
```

File: src/viur/core/skeleton.py [hunk 5]
```diff
@@ -1061,75 +1055,75 @@ def fixDotNames(entity):
...
                 else:
-                    tmp = set(oldBlobLockObj["old_blob_references"] + [x for x in removedBlobs])
-                    oldBlobLockObj["old_blob_references"] = [x for x in (tmp - blobList)]
-                oldBlobLockObj["has_old_blob_references"] = \
-                    oldBlobLockObj["old_blob_references"] is not None \
-                    and len(oldBlobLockObj["old_blob_references"]) > 0
-                oldBlobLockObj["is_stale"] = False
-                db.Put(oldBlobLockObj)
+                    old_blob_refs = set(old_blob_lock_obj["old_blob_references"])
+                    old_blob_refs += set([removed_blob for removed_blob in removed_blobs])  # Add removed blobs
```
Comment (sveneberth) [inline src/viur/core/skeleton.py:1079]:
```suggestion
                    old_blob_refs.update(removed_blobs)  # Add removed blobs
```

File: src/viur/core/skeleton.py [hunk 6]
```diff
@@ -1061,75 +1055,75 @@ def fixDotNames(entity):
...
+                    old_blob_refs -= blob_list  # Remove active blobs
+                    old_blob_lock_obj["old_blob_references"] = [old_blob_ref for old_blob_ref in old_blob_refs]
```
Comment (sveneberth) [inline src/viur/core/skeleton.py:1081]:
```suggestion
                    old_blob_lock_obj["old_blob_references"] = list(old_blob_refs)
```

**Review by sveneberth — CHANGES_REQUESTED** [review]

**Review by phorward — CHANGES_REQUESTED** [review]
@ArneGudermann can you resolve [the comment](https://github.com/viur-framework/viur-core/pull/973#discussion_r1459105588) from @sveneberth?

File: src/viur/core/skeleton.py [hunk 7]
```diff
@@ -873,185 +873,179 @@ def toDB(cls, skelValues: SkeletonInstance, update_relations: bool = True, **kwa
...
+            db_key = write_skel["key"]
+            skel = write_skel.skeletonCls()
 
+            blob_list = set()
+            change_list = []
+            old_copy = {}
             # Load the current values from Datastore or create a new, empty db.Entity
-            if not dbKey:
+            is_add = not bool(db_key)
+            if is_add:
```
Comment (phorward) [inline src/viur/core/skeleton.py:893]: Please use modern syntax!
```suggestion
            if is_add := not bool(db_key):
```

**Review by phorward — CHANGES_REQUESTED** [review]
Almost finished! :smile:

**Review by phorward — DISMISSED** [review]

**Comment by sveneberth** [comment]
@ArneGudermann Can you resolve the merge conflict?

File: src/viur/core/skeleton.py [hunk 8]
```diff
@@ -870,182 +870,183 @@ def toDB(cls, skelValues: SkeletonInstance, update_relations: bool = True, **kwa
...
-                            db.Put(newLockObj)
-                        if newLockValue in oldUniqueValues:
-                            oldUniqueValues.remove(newLockValue)
-                    dbObj["viur"][f"{key}_uniqueIndexValue"] = newUniqueValues
+                            lock_obj = db.Entity(new_lock_key)
+                            lock_obj["references"] = db_obj.key.id_or_name
+                            db.Put(lock_obj)
+                        if new_lock_value in old_unique_values:
+                            old_unique_values.remove(new_lock_value)
+                    db_obj["viur"][f"{key}_uniqueIndexValue"] = new_unique_values
```
Comment (sveneberth) [inline src/viur/core/skeleton.py:962]:
```
   db_obj["viur"][f"{db_key}_uniqueIndexValue"] = new_unique_values
NameError: free variable 'key' referenced before assignment in enclosing scope
```

```suggestion
                    db_obj["viur"][f"{db_key}_uniqueIndexValue"] = new_unique_values
```

**Review by sveneberth — CHANGES_REQUESTED** [review]

File: src/viur/core/skeleton.py [hunk 9]
```diff
@@ -870,182 +870,183 @@ def toDB(cls, skelValues: SkeletonInstance, update_relations: bool = True, **kwa
...
-                            db.Put(newLockObj)
-                        if newLockValue in oldUniqueValues:
-                            oldUniqueValues.remove(newLockValue)
-                    dbObj["viur"][f"{key}_uniqueIndexValue"] = newUniqueValues
+                            lock_obj = db.Entity(new_lock_key)
+                            lock_obj["references"] = db_obj.key.id_or_name
+                            db.Put(lock_obj)
+                        if new_lock_value in old_unique_values:
+                            old_unique_values.remove(new_lock_value)
+                    db_obj["viur"][f"{db_key}_uniqueIndexValue"] = new_unique_values
```
Comment (sveneberth) [inline src/viur/core/skeleton.py:962]: Ahh, @phorward you were to fast. `bone_name` would be correct. _db_key_ was simply too straightforward

```suggestion
                    db_obj["viur"][f"{bone_name}_uniqueIndexValue"] = new_unique_values
```

**Review by sveneberth — APPROVED** [review]

**Review by phorward — APPROVED** [review]

**Pull request merged**